tests/RandomTest.php

```diff
@@ -71,6 +75,83 @@ protected function init(): void
     }
 }
 
+trait ModelSoftDeleteTrait
+{
+    protected function initSoftDelete(): void
+    {
+        $this->addField('is_deleted', ['type' => 'boolean', 'nullable' => false, 'default' => false]);
+        $this->addCondition('is_deleted', false);
+        $this->onHook(Model::HOOK_BEFORE_DELETE, function (Model $entity) {
+            $softDeleteController = new ControllerSoftDelete();
+            $softDeleteController->softDelete($entity);
+
+            $entity->hook(Model::HOOK_AFTER_DELETE);
+            $entity->breakHook(false); // this will cancel original Model::delete()
```

Comment:
an alternative could be to replace the persistence during update so the delete hook does not need to be broken

Reply:
Both solutions are internal so we can go ahead in this way and add this feature.

Reply:
Not really, such changes can have effect on overrided delete method or user hooks, very important decision that needs literaly hours of very fine reasoning to decide this correctly.